scripts/gb_model/calc_interconnector_bid_offer_profile.py

```diff
@@ -0,0 +1,197 @@
+#
+# SPDX-License-Identifier: MIT
+
+"""
+Calculate interconnector bids and offer prices
+"""
+
+import logging
+from pathlib import Path
+
+import pandas as pd
+import pypsa
+
+from scripts._helpers import configure_logging, set_scenario_config
+from scripts.gb_model._helpers import filter_interconnectors, marginal_costs_bus
+
+logger = logging.getLogger(__name__)
+
+
+def _calculate_interconnector_loss(
+    unconstrained_result: pypsa.Network, interconnectors: pd.DataFrame
+) -> pd.DataFrame:
+    """
+    Calculate capacity weighted average of interconnector losses at each foreign market
+
+    Parameters
+    ----------
+    unconstrained_result: pypsa.Network
+        Result of the unconstrained optimization
+    interconnectors: pd.DataFrame
+        Dataframe of interconnectors and their pypsa parameters
+    """
+    interconnector_grouping = (
+        interconnectors.index.to_series().groupby(interconnectors["bus1"]).agg(list)
+    )
+    loss_profile = pd.DataFrame(
+        index=unconstrained_result.snapshots, columns=interconnector_grouping.index
+    )
+    for bus in interconnector_grouping.index:
+        group = interconnector_grouping.loc[bus]
+        # Capacity weighted average of interconnectors connected to each EU node
+        loss_profile[bus] = (
+            (
+                unconstrained_result.links_t.p1[group]
+                - unconstrained_result.links_t.p0[group] * -1
+            )
+            .abs()  # interconnector loss between GB and EU node
+            .mul(interconnectors.loc[group].p_nom)
+            .sum(axis=1)
+            .div(interconnectors.loc[group].p_nom.sum())
+        )
+
+    logger.info("Calculated the interconnector loss profile")
+    return loss_profile
+
+
+def calc_interconnector_bids_and_offers(
+    interconnectors: pd.DataFrame,
+    EU_marginal_gen_profile: pd.DataFrame,
+    interconnector_fee_profile: pd.DataFrame,
+    loss_profile: pd.DataFrame,
+) -> tuple[pd.DataFrame]:
+    """
+    Calculate the bids and offers for import, export and floating condition of each interconnector
+
+    Parameters
+    ----------
+        interconnectors: pd.DataFrame
+            Dataframe of interconnectors from the pypsa network.links
+        EU_marginal_gen_profile: pd.DataFrame
+            bid / offer for the marginal generator at each EU bus
+        interconnector_fee_profile: pd.DataFrame
+            Spread in marginal prices for each interconnector
+        loss_profile: pd.DataFrame
+            Weighted capacity average of losses of interconnectors connected to each EU bus
+    """
+    (
+        import_bid_profile,
+        import_offer_profile,
+        float_import_profile,
+        export_bid_profile,
+        export_offer_profile,
+        float_export_profile,
+    ) = (
+        pd.DataFrame(
+            index=interconnector_fee_profile.index, columns=interconnectors.index
+        )
+        for _ in range(6)
+    )
```

Comment:
This would be easier to track and store if collapsed into a single table with a multiindex